core/vm/contracts.go

```diff
@@ -502,3 +526,493 @@ func (c *blake2F) Run(input []byte) ([]byte, error) {
+
+	// Encode the G2 point into 256 bytes
+	return encodeBLS12381G2Point(g, r), nil
+}
+
+// bls12381G2MultiExp implements EIP-2537 G2MultiExp precompile.
+type bls12381G2MultiExp struct{}
+
+// RequiredGas returns the gas required to execute the pre-compiled contract.
+func (c *bls12381G2MultiExp) RequiredGas(input []byte) uint64 {
+	k := len(input) / 160
+	maxDiscountLen := len(params.Bls12381MultiExpDiscountTable)
+	if k >= maxDiscountLen {
+		k = maxDiscountLen
+	}
+	discount := params.Bls12381MultiExpDiscountTable[k]
+	return (uint64(k) * params.Bls12381G2MulGas * discount) / 1000
+}
+
+func (c *bls12381G2MultiExp) Run(input []byte) ([]byte, error) {
+	// Implements EIP-2537 G2MultiExp precompile logic
+	// > G2 multiplication call expects `288*k` bytes as an input that is interpreted as byte concatenation of `k` slices each of them being a byte concatenation of encoding of G2 point (`256` bytes) and encoding of a scalar value (`32` bytes).
+	// > Output is an encoding of multiexponentiation operation result - single G2 point (`256` bytes).
+	k := len(input) / 288
+	if len(input) != k*288 || k == 0 {
+		return nil, errBLS12381InvalidInputLength
+	}
+	var err error
+	points := make([]*bls12381.PointG2, k)
+	scalars := make([]*big.Int, k)
+
+	// Initialize G2
+	g := bls12381.NewG2()
+
+	// Decode point scalar pairs
+	for i := 0; i < k; i++ {
+		off := 288 * i
+		t0, t1, t2 := off, off+256, off+288
+		// Decode G1 point
+		if points[i], err = decodeBLS12381G2Point(g, input[t0:t1]); err != nil {
+			return nil, err
+		}
+		// Decode scalar value
+		scalars[i] = new(big.Int).SetBytes(input[t1:t2])
+	}
+
+	// Compute r = e_0 * p_0 + e_0 * p_0 + ... + e_(k-1) * p_(k-1)
+	r := g.New()
+	_, _ = g.MultiExp(r, points, scalars)
+
+	// Encode the G2 point to 256 bytes.
+	return encodeBLS12381G2Point(g, r), nil
+}
+
+// bls12381Pairing implements EIP-2537 Pairing precompile.
+type bls12381Pairing struct{}
+
+// RequiredGas returns the gas required to execute the pre-compiled contract.
+func (c *bls12381Pairing) RequiredGas(input []byte) uint64 {
+	return params.Bls12381PairingBaseGas + uint64(len(input)/384)*params.Bls12381PairingPerPairGas
+}
+
+func (c *bls12381Pairing) Run(input []byte) ([]byte, error) {
+	// Implements EIP-2537 Pairing precompile logic.
+	// > Pairing call expects `384*k` bytes as an inputs that is interpreted as byte concatenation of `k` slices. Each slice has the following structure:
+	// > - `128` bytes of G1 point encoding
+	// > - `256` bytes of G2 point encoding
+	// > Output is a `32` bytes where last single byte is `0x01` if pairing result is equal to multiplicative identity in a pairing target field and `0x00` otherwise
+	// > (which is equivalent of Big Endian encoding of Solidity values `uint256(1)` and `uin256(0)` respectively).
+	L := 384
+	k := len(input) / L
+	if len(input) != k*L || k == 0 {
+		return nil, errBLS12381InvalidInputLength
+	}
+
+	// Initialize BLS12-381 pairing engine
+	e := bls12381.NewPairingEngine()
```

Comment:
Is it possible to cache and use `e.reset()` it before usage? 
Creating the Engine from scratch uses ~10KB that have to be allocated and garbage collected every call, see:
https://play.golang.org/p/PJ8iub-DG1l


Do you know if `cfgArch()` gets compiled out by go?
Its called in `NewEngine` (https://github.com/kilic/bls12-381/blob/f68242f46180fdfcc9229c8bf29d64adb4984cdb/pairing.go#L24) as well as in `NewG1()` and `NewG2()`

Reply:
> Is it possible to cache and use `e.reset()` it before usage?
> Creating the Engine from scratch uses ~10KB that have to be allocated and garbage collected every call, see:
> https://play.golang.org/p/PJ8iub-DG1l

The reason for initializing new engine in every call is that underlying library is not suitable for concurrent operations with the same instance since an instance of pairing engine and G1 and G2 has its own allocated temporary variables. If we are sure that precompiles in EIP2537 will not concurrently _in any case_ then we can definitely initialize pairing engine once.

> Do you know if `cfgArch()` gets compiled out by go?
> Its called in `NewEngine` (https://github.com/kilic/bls12-381/blob/f68242f46180fdfcc9229c8bf29d64adb4984cdb/pairing.go#L24) as well as in `NewG1()` and `NewG2()`

We can supply a build tag to target specific extensions in compile time. However in my knowledge I don’t know how compiler can detect the existence of specific extension in a x86 machine. Namely, here it’s [ADX extension](https://en.wikipedia.org/wiki/Intel_ADX).

Reply:
> If we are sure that precompiles in EIP2537 will not concurrently in any case then we can definitely initialize pairing engine once.

They will definitely run concurrently (the RPC api may use them, and the precacher, and maybe also the txpool)